pkg/sql/contention/txnidcache/txn_id_cache.go

```diff
@@ -0,0 +1,178 @@
+// Copyright 2021 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package txnidcache
+
+import (
+	"context"
+
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
+	"github.com/cockroachdb/cockroach/pkg/util/encoding"
+	"github.com/cockroachdb/cockroach/pkg/util/stop"
+	"github.com/cockroachdb/cockroach/pkg/util/uuid"
+)
+
+const (
+	shardCount  = 256
+	channelSize = 128
+)
+
+// Cache stores the mapping from the Transaction IDs (UUID) of recently
+// executed transactions to their corresponding Transaction Fingerprint ID (uint64).
+// The size of Cache is controlled via sql.contention.txn_id_cache.max_size
+// cluster setting, and it follows FIFO eviction policy once the cache size
+// reaches the limit defined by the cluster setting.
+//
+// Cache's overall architecture is as follows:
+//   +------------------------------------+
+//   | connExecutor ---------*            |
+//   |      +-------------+  | writes to  |
+//   |      | Writer      |<-*            |
+//   |      +-----+-------+               |
+//   +------------|-----------------------+
+//                |
+//               when full, Writer flushes into a channel.
+//                |
+//                v
+//              channel
+//                ^
+//                |
+//              Cache runs a goroutine that polls the channel --*
+//                |  and sends the message to its               |
+//                |  corresponding shard                        |
+//                |               *-----------------------------*
+//   +----------------------------|---+
```

Comment:
It feels to me like the connExecutor should just write straight to a shard. The shard is sharded and the channel is not. What would happen if you simplified it such that when a shard is full, you send the shard itself on the channel to flush and swap it out with a new shard? You can then sync.Pool the shards